### PR TITLE
Hw tests

### PR DESCRIPTION
**Pull request by baali**
This is combined PR against issue #77 for following tests:
- [x] LED - Toggling the pin state for GPIO 198 `hwtests.sh`
- [x] Wifi (via USB 1) - `wifi_test.sh` PR #78 
  - [x] Scan
  - [x] Connect - Needs `SSID_NAME`, `SSID_PASSWORD` environment variable to connect to SSID
  - [x] Send/Receive data - We are using curl command to fetch HTML file and confirm file size.
  - [x] Disconnect
- [x] Bluetooth (via UART) - `bluetooth_test.sh` PR #89 
  - [x] Scan
  - [x] Pair
  - [x] Send/Receive data - Just fetching `info` of scanned device to confirm data exchange. 
  - [x] Disconnect
- [x] USB 2 (external port, empty by default) 
  - [x] Use an additional bluetooth or wifi dongle to perform a similar test as above - Above test will use all devices available irrespective of USB port.

Recipe file can be used to copy respective shell scripts to the image.

### CODE REVIEW

File: meta-senic/recipes-core/initscripts/initscripts_1.0.bbappend [hunk 1]
```diff
@@ -1,22 +1,19 @@
...
-
-# Tests needed only on debug release
-SRC_URI += "file://esd_test.sh"
+DEPENDS = "qrencode "
+SRC_URI += "file://hwtests.sh \
+        file://bluetooth_test.sh \
+        file://wifi_test.sh \
+        "
 
 do_install_append() {
```
Comment (kelada) [inline meta-senic/recipes-core/initscripts/initscripts_1.0.bbappend]: Remove white space

File: meta-senic/recipes-core/initscripts/initscripts_1.0.bbappend [hunk 2]
```diff
@@ -1,22 +1,19 @@
...
-  install -m 0755 ${WORKDIR}/networkmanager ${D}${sysconfdir}/init.d/networkmanager
-
-  # Add initscripts to runlevel 5 (default)
-  update-rc.d -r ${D} senic_update_motd start 98 5 .
-  update-rc.d -r ${D} networkmanager start 91 5 .
+  # Install shell scripts into /usr/bin/
+  install -d ${D}${bindir}
+  install -m 0755 ${WORKDIR}/hwtests.sh ${D}${bindir}/
+  install -m 0755 ${WORKDIR}/bluetooth_test.sh ${D}${bindir}/
+  install -m 0755 ${WORKDIR}/wifi_test.sh ${D}${bindir}/
```
Comment (kelada) [inline meta-senic/recipes-core/initscripts/initscripts_1.0.bbappend]: Remove white space

**Comment by kelada** [comment]
This looks pretty good @baali, I have not tested this yet. My only critics would be:
- [ ] Re-add the indentation in `initscripts_1.0.bbappend`
  - Lines 5-7 & 10-14 should be indented
- [ ] I challenge the validity of the wifi test in so far as downloading the webpage
  - Perhaps being able to parse data download from the website makes more sense, like reading the GET response headers and/or status code.

As for the LED issue, you should talk to @rudolfs about this.

**Comment by baali** [comment]
Thank you @TheMeaningfulEngineer for the inputs. Yes I used your commits on `senic_hub/backend/nuimo_setup.py` to sort out issue with unresponsive dongle. I will add logic to completely restart the module also. 
About testing connection to Nuimo, at the moment logic I am following is to scan for any BLE device and connect to it using `hcitool`. I will try to use `nuimoctl` as you have suggested and shift to using this instead. 
:+1:

**Comment by baali** [comment]
Yes jonathan, my bad, I was referring to someone else.

Hello @TheMeaningfulEngineer I tried to use `nuimoctl --discover --adapter hci0` but I think its more tuned to just work with Nuimo, for me, when I try this there are no results returned and I don't have a Nuimo either. The current tools, `hcitool` and `hciconfig` are generic ones which I am able to use to get basic communication with any Bluetooth device.

Should we tailor our tests specific for Nuimo device or keep them generic like we have right now? @TheMeaningfulEngineer @kelada.

**Comment by baali** [comment]
@kelada regarding the comment on using `curl` to parse the header, this can work:

    $ curl -I -s https://www.senic.com/en/ | grep server | cut -d' ' -f2
    Netlify

[Netlify](https://www.netlify.com/) is the CDN [we](https://www.senic.com/en/) are using I think.

**Comment by TheMeaningfulEngineer** [comment]
> Should we tailor our tests specific for Nuimo device or keep them generic like we have right now? 

What I didn't get is what are the underlying assumptions we're testing the generic parts on.
For what I understood it for now it's: 
"There will be something in the BT spectrum always, and we're basing the testing on detecting that?"

Is this correct?

**Comment by baali** [comment]
> "There will be something in the BT spectrum always, and we're basing the testing on detecting that?"

Yes, with any/some Bluetooth device in "range", test should be able to scan them, try to `connect` to them, query their `info` and then `disconnect`. These tests/operations, didn't use any more configurations like pairing using pass keys, exchanging particular device command.  And if these work, inferring that our BT dongle is working Okay.

**Comment by TheMeaningfulEngineer** [comment]
Alright.
I would then say let's stick with until proven false :)
Just please extend it with a print statement stating the assumption.
In a scenario where there is no BT in the air ( :D ) we'd end with a false
hardware detection.

**Comment by kelada** [comment]
@baali could you change the following:

https://github.com/baali/senic-os/blob/9490e80e91613b10122b832b08f2c3a9de792647/meta-senic/recipes-core/initscripts/initscripts/hwtests.sh#L20-L29

to:
```
# Test for LED, Manual prompt for now, given that GPIO6, being part of
# device tree, is locked from userspace.
LED_TEST="F"
while true; do
    read -p "Is the status LED On? y/n: " -i "y" led_status
    case $led_status in
        [Yy]* )
            LED_TEST="P";;
        [Nn]* )
            echo " Failed the LED test";;
        * ) echo " Please answer y/n.";;
    esac
done
```

Please test this, but basically it makes sure a user actually presses y/n and doesn't press something accidentally.

**Comment by kelada** [comment]
Use the following QR code command: `qrencode -t ANSI256 -l H`

**Comment by TheMeaningfulEngineer** [comment]
Solved with this: 
https://github.com/getsenic/senic-os/commit/106dc293aad64821539ba306b2b54065517f5a0b

**Pull request closed without merging**